CMakeLists.txt

```diff
@@ -13,7 +13,7 @@ foreach(p
   endif()
 endforeach()
 
-project (benchmark CXX)
+project (benchmark VERSION 1.5.2 LANGUAGES CXX)
```

Comment:
While this will work, i'm not okay with this.
This will falsely claim that every build of every commit is the released version.
Such a change (this particular line) should be the last commit before release,
and reverted immediately after.

IOW, this should be documented in "how to release benchmark", but NOT changed now.

Reply:
Well, it's a question of what you consider "claim". The cmake version for this project will be set to 1.5.2, which yes, is incorrect (or at least incomplete) after additional commits are added. But before, that version wasn't set at all, so it doesn't really have any impact in real life, it's not even printed. 

What already happens right now is that if you install master as-is, in its install directory, the config file will contain a version of  `1.5.2`. `git describe` will say `1.5.2-<commit>`, and `-dirty` may be added if there are changes in the work tree, but that info is only printed. What's recorded in the install is just `1.5.2`, and yes, that's not correct, but it's already the case.

But anyway, being aware of this issue, the first solution I proposed in #1046 would not add a version  to `project()` in master, so it would continue to determine the version based on `git describe` (with still the same issue that what's recorded is just the info from the tag, even if we're ahead of that tag). But it wouldn't change existing behavior at all. It would, however, require to make a branch at release time and add the particular release version to the `project()` command at that point, so that a tarball made from that release branch will install benchmark with the proper version info. (It's not really required to necessarily a branch, it could also be another part of the release process to add the version info `project()` before making the tarball, but in that case the tarball wouldn't correspond exactly to the tagged version).

Reply:
> It would, however, require to make a branch at release time and add the particular release version to the `project()` command at that point, so that a tarball made from that release branch will install benchmark with the proper version info.

Why would it require a branch?

Reply:
I'm not a fan of creating a branch for a release, and i'm not a fan of adding a commit just to reverse it after the release. I think having the project include a "base version" and having git describe override that with a more detailed version is appropriate.

Reply:
A release branch would allow you to add a commit that does
```diff
- project (benchmark CXX)
+ project (benchmark VERSION 1.5.2 LANGUAGES CXX)
```
outside of `master`.

The alternatives that I can see are:
* add the commit above to`master`, tag the release, revert it.
* don't ever commit the change above, but apply it locally and make the release tarball.

The version of the release has to become part of what's in the tarball in one way or other (it doesn't have to come from the `project()` command, either, but that's the standard way of specifying a version). I don't really care which way, but it'd be nice to be able to use a release via `find_package(benchmark 1.5.2)`.

FWIW, the reason that I started looking at this is that `benchmark=1.5.2` installed via conda thinks it's version `0.0.0`.

Reply:
i think the only way to do this that works is to have the change as it is in this commit but with additional step added to https://github.com/google/benchmark/blob/master/docs/releasing.md to change the project here.

that way, the tarball will be correct (it's generated by github) and local changes will still reflect the tag + optional dirty flag.

Reply:
@dominichamon can you explain why the solution of adding this just before the tag,
and reverting right afterwards is so repulsive to you?
FWIW this would not really be a new invention by this project.

Reply:
It's too easy to forget. Why is having the project version permanently set such a problem?

Reply:
Because for non-releases (i.e. git snapshots) it's set to a wrong version :[
Unfortunately i'm very much aware how complicated all this versioning stuff is,
especially if we account for automatic github tarballs,
so i really would suggest having //some//
How about:
* for non-releases, the version is hardcoded to `vlast.release+git` string
* just before the tag, it is changed to `vnew.release` string
* right afterwards it is updated to `vnew.release+git` string

Reply:
how would it be wrong? if git version responds something then we override the project version with that (is that maybe not possible?).

ie, we should:

1. set the version to the last release in the project statement
2. call the script to calculate the git version and set that (or at least output it to the cmake logs... we only really need it for debugging and triage right?).
3. on release, update the project version to the new tag before tagging.

if we were on cmake 3.15 then we could use cmake_project_include_before to calculate a git version and return the "current version" if none is set, i think. then the release step would be to change the default version in that script.

Reply:
> how would it be wrong?

> if git version responds something then we override the project version with that (is that maybe not possible?).
Unfortunately, that only works if the whole `.git` is there.

E.g. what's the version of https://github.com/google/benchmark/archive/master.zip ?
It's really ugly/messy to handle in full generality, that is why i'm suggesting pre-tag/post-tag commits.

> ie, we should:
> 
> 1. set the version to the last release in the project statement

> 2. call the script to calculate the git version and set that (or at least output it to the cmake logs... we only really need it for debugging and triage right?).

I guess. Packagers know what sources they've downloaded, and can and should use appropriate version in any case.

> 3. on release, update the project version to the new tag before tagging.
> 
> if we were on cmake 3.15 then we could use cmake_project_include_before to calculate a git version and return the "current version" if none is set, i think. then the release step would be to change the default version in that script.

Reply:
> > how would it be wrong?
> 
> > if git version responds something then we override the project version with that (is that maybe not possible?).
> 
> Unfortunately, that only works if the whole `.git` is there.
> E.g. what's the version of https://github.com/google/benchmark/archive/master.zip ?

oh i hadn't thought of that one. so that would show the version as the previous release but all we'd know it was at a commit past that. and we don't have a dirty flag for that case. got it.

i  wonder how much that's a problem really. if someone downloads the master.zip and installs it and then it doesn't work, we can suggest they get a tagged release (most would) or go from source (others would do that), which would both have appropriate versions set.

> It's really ugly/messy to handle in full generality, that is why i'm suggesting pre-tag/post-tag commits.
> 
> > ie, we should:
> > 
> > 1. set the version to the last release in the project statement
> 
> > 1. call the script to calculate the git version and set that (or at least output it to the cmake logs... we only really need it for debugging and triage right?).
> 
> I guess. Packagers know what sources they've downloaded, and can and should use appropriate version in any case.
> 
> > 1. on release, update the project version to the new tag before tagging.
> > 
> > if we were on cmake 3.15 then we could use cmake_project_include_before to calculate a git version and return the "current version" if none is set, i think. then the release step would be to change the default version in that script.

Reply:
Currently, benchmark does the following when installing from a git repo (this behavior is unchanged by this PR):
```
-- git Version: v1.5.2-7c7e3a75
-- Version: 1.5.2
```
When installed, the installed `benchmarkConfigVersion.cmake` has
```
set(PACKAGE_VERSION "1.5.2")
```

So the installed version does not record that it's from a version past the 1.5.2 release. Maybe this should be changed (but the next question then would be how you'd want it changed -- `1.5.2-dev`? I'm not sure whether cmake would consider that to be compatible when `1.5.2` is requested. This would be a (possibly breaking) change from current behavior. If you do want to preserve that information,  maybe a better option would be to generate a `benchmark.h` that contains `#define BENCHMARK_VERSION "v1.5.2-7c7e3a75"`)

This PR only changes behavior when `.git` is not available.

Instead of current behavior
```
-- git Version: v0.0.0
-- Version: 0.0.0 # <- gets recorded when installed
```
It will use the version from the project command (this would happen both with the 1.5.2 release, and a later `master.zip`.)
```
-- git Version: v0.0.0
-- Version: 1.5.2 # <- gets recorded when installed
```

I don't think it's unreasonable to do:
* last commit before release is to put `VERSION 1.5.3` into the `project()` command
* tag the release
* next commit changes it to `VERSION 1.5.3-dev`

Though the way things currently are, if you install from the master branch then, you'll get a recorded `1.5.3` version if it was installed from a repo with `.git`, and you'll get a recorded `1.5.3-dev` if installed from `master.zip`, which doesn't really seem desirable.

Reply:
i like the idea of having the version be set as part of the release and then _updated_ to something between the release and the next release (1.5.3.1 or 1.5.3-dev). if you get the master.zip and install it then you'd need to use that tag to match it, which i think is fair. you're explicitly looking for the dev (non official release) version.